cds-snc/ircc-rescheduler
# Replace Apollo Link State for all of our user-submitted values

We want to support a no-JS flow through our app -- that is, we want our app to work for clients who aren't running or can't run JavaScript.

Among other things, this means we need to:
- replace our caching layer because Apollo Link State only saves values to the client (whereas we need to be able to decide what HTML to send back from the server)
- handle the default form behaviour of submitting a request to the server with a bunch of query parameters

**This PR** 

- replaces Apollo Link state for our user-submitted values (the registration form and the calendar page).
  - It uses [React context](https://reactjs.org/docs/context.html) as a global state that can be passed around to nested components
  - It saves the context as a browser cookie and re-reads it later to support persistence between hard refreshes or a no-JS flow through the app.
- removes most of the Apollo Link State config from `utils/createApolloClient.js`
- removes the `LinkStateDisplay` component that was only useful for testing local cached values

**This PR does not**
- Redirect after successfully submitting your values on the registration page without JS. 
  - If you submit bad data, you will get validation messages
  - If you submit good data, you will just be returned to the page
- Add a straightforward way for users to clear out their session (ie, this should work with the 'cancel request' button)
- Make any consistent decisions about how long to save cookies for
- Remove all uses of Apollo Link State: we still have the language switcher to convert
- Make the calendar page work without JS. That is another story.
- Add any new tests. This totally sucks: this is testable code, but we're pressed for time.

**To test this PR**

Install new dependencies, and then build and run the app.
You should be able to do all of the things you can currently do: make it back and forth through the app with your values being saved between pages.

New things you can do:
- on the `/register` page, you can hit "continue" without JS running on the page
  - any values that pass validation will be saved
  - any values that fail validation will cause error messages to show up when the page reloads
- from any point in the app, you can do a hard refresh and not lose your previously saved values

**To learn more**

Reading the diff for this PR is terrible because there's a lot of highlighted line diffs that came about from un-indenting the nesting in various places. The actual code changes are hard to find amongst the whitespace changes.

A simplified version of all this code can be found at my [after-no-js](https://github.com/pcraig3/after-no-js) repo where I trialled all this stuff.
- read [REQUEST_LIFECYCLE.md](https://github.com/pcraig3/after-no-js/blob/master/REQUEST_LIFECYCLE.md) for an overview
- build the project and click around 
  - if you uncomment the console statements, you will see when it sets and reads cookies
- ask me about it
